### PR TITLE
Added new features (write, buffer, commit) for the ReadWrite channel …

### PR DESCRIPTION
**Pull request by eatondan**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-custom-device-inline-async-api/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Updated the Read/Write channel type to have more write based features. Added Write, Buffer, Commit functions that match the already available Write channel options.

### Why should this Pull Request be merged?

New features are required by custom device developers at our customers.

### What testing has been done?

Features were tested by me in a custom device as well as a customer in a different custom device.

### CODE REVIEW

**Comment by rtzoeller** [comment]
Thanks @eatondan for the pull request! Please complete the pull request template so we can better understand the changes it includes and the motivation behind them.

**Comment by eatondan** [comment]
@rtzoeller  
Updated. Thanks!

**Comment by csjall** [comment]
I am confused by the changes in ReadWrite.ProcessData.Read.vi. 
image

- [ ] In the previous revision we used the greater than comparison, but now we are using not equal to. Why did the comparison change?
- [ ] The logic for the "Ignore Reset Values?" terminal seems reversed. In the true case we are referring to the reset values. I would not expect to reference the reset values when we are ignoring them.

In ReadWrite.ProcessData.vi 
image
- [ ] This VI still has the old implementation of calculating "Detect?" array. Should this VI be updated to match the implementation in Read.vi.

**Comment by eatondan** [comment]
The original version of the read/write channel type was designed as a trigger. So, the write (or reset) was always 0. Therefore, a greater than was an easy comparison. Part of the update is to allow the user to write values other than zero. This means the comparison had to be updated. 

The Ignore Reset Value allows the original trigger like functionality by ignoring the reset. Again, the original trigger method would go from 0 to 1 and then reset from 1 to 0. The greater than implementation would trigger on the 0 to 1 and not the 1 to 0. This new triggering method also the user to either trigger the original way by ignoring the 1 to 0 (ignore reset value) or trigger on both.

The ProcessData uses the old implementation (greater than) because it doesn't because it doesn't allow the user to explicitly write. It is just a read and reset to 0 (trigger function). The new function is a separate read that can be used with a separate write.

File: Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Access Type Classes/ReadWrite/ReadWrite.lvclass [hunk 1]
```diff
@@ -6,9 +6,9 @@
...
 	<Property Name="NI.Lib.SourceVersion" Type="Int">352354304</Property>
 	<Property Name="NI.Lib.Version" Type="Str">1.0.0.0</Property>
-	<Property Name="NI.LV.All.SourceOnly" Type="Bool">true</Property>
+	<Property Name="NI.LV.All.SourceOnly" Type="Bool">false</Property>
```
Comment (buckd) [inline Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Access Type Classes/ReadWrite/ReadWrite.lvclass]: Source should still be separated.

File: Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Data Access Engine/Data Access Engine.lvclass [hunk 2]
```diff
@@ -6,10 +6,10 @@
...
 	<Property Name="NI.Lib.SourceVersion" Type="Int">352354304</Property>
 	<Property Name="NI.Lib.Version" Type="Str">1.0.0.0</Property>
-	<Property Name="NI.LV.All.SourceOnly" Type="Bool">true</Property>
+	<Property Name="NI.LV.All.SourceOnly" Type="Bool">false</Property>
```
Comment (buckd) [inline Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Data Access Engine/Data Access Engine.lvclass]: Source should still be separated.

**Review by buckd — CHANGES_REQUESTED** [review]
- [ ] All the VIs you created/modified had their Source Only setting changed. All of these VIs should keep their source separated.
- [ ] Is there a reason `ReadWrite.lvclass:ReadWrite.ProcessData.Write.vi` re-implements the `Buffer` and `Commit` VI logic instead of just calling those subVIs?

File: Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Access Type Classes/ReadWrite/ReadWrite.lvclass (hunk 1 above)
Comment (eatondan) [reply Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Access Type Classes/ReadWrite/ReadWrite.lvclass]: Fixed and recommitted.

File: Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Data Access Engine/Data Access Engine.lvclass (hunk 2 above)
Comment (eatondan) [reply Source/vi.lib/NI/NIVS Inline Async API/_Data Access Engine/Data Access Engine/Data Access Engine.lvclass]: Fixed and recommitted.

**Comment by eatondan** [comment]
On "Is there a reason ReadWrite.lvclass:ReadWrite.ProcessData.Write.vi re-implements the Buffer and Commit VI logic instead of just calling those subVIs?". The classes are not compatible. You can't directly re-use and re-wire.

**Comment by buckd** [comment]
> The classes are not compatible. You can't directly re-use and re-wire.

The VIs are in the same class. This code looks identical:
image

If one part of the implementation has to change, we now have to change it in two places. By reusing the subVIs, there is only one location that must be updated.
image

**Comment by csjall** [comment]
It still feels like the Boolean logic is inverted. If "Ignore Reset Values?" is true, why are we evaluating the reset values when detecting a change? In my mind the true/false case should be swapped.

image

**Comment by eatondan** [comment]
On the buffer and commit, I see what you mean now.  I can make that change.

**Comment by eatondan** [comment]
On the inverted logic, we need to use the reset values to ignore them. The logic should be don't allow (ignore) any triggering (detection) when the written values match the reset values. If you want to change the wording of the input, I am happy to do that. Just let me know what makes more sense.

Thank you for all the help in this! I really appreciate it!

**Comment by csjall** [comment]
Go ahead and change the name of the terminal to be concise yet precise.

**Comment by eatondan** [comment]
I could make it "Do Not Detect Reset". Thoughts?

**Comment by csjall** [comment]
I prefer affirmative names like "Detect Reset Value" or "Ignore Reset Value" as opposed to "Do Not Detect Reset Value". I would rather flip the case structure than add the word "Do Not" as a prefix.

@tannerblair and I were talking about this. It seems the goal is to detect/ignore a reset value condition. Could the Reset Value SubVI set a boolean flag when the values is reset. This boolean flag can be evaluated and cleared the next time we Read the values. What is the benefit of cheecking all the reset values for a trigger condition as opposed to checking a single boolean?

**Comment by eatondan** [comment]
I agree on the affirmative names. I should have done that initially. However, I really don't want to switch the cases because it would break released code. It should have been Detect Reset and cases flipped. Now that it is released, it should be left as is or changed to Do Not Detect Reset. Thoughts?

On the reset flag. That would be something you need to save per channel. Each array element can be individually reset. So, you would end up with a reset detection array. It isn't much different than what I already have. The checking is the easiest, most robust way to handle it.

**Comment by csjall** [comment]
How is this released already. I thought the Ignore Reset Value is a new terminal.

**Comment by eatondan** [comment]
I didn't know you branched my API. A customer needed new features. I added them to my branch. Tanner let me know about this version of the branch last week and asked me to merge them and deprecate mine. That is what I am trying to do.

**Comment by csjall** [comment]
No worries. Go ahead and submit a pull request with name change.

**Comment by eatondan** [comment]
Add updated the repo based on our discussion. Do I need to close this pull request and do another one?

**Comment by csjall** [comment]
You can keep this pull request. The pull request is automatically updated when you submit changes to your master branch.

**Review by buckd — COMMENTED** [review]
There are still 3 files that do not have source separated. Other than this, everything looks good.
- [ ] `Data Access Engine.lvclass`
- [ ] `ReadWrite.lvclass`
- [ ] `Data Access Engine.lvclass:Data.Process.Execute.vi`

**Comment by eatondan** [comment]
I apologize. Those are now updated as well.

**Comment by eatondan** [comment]
I apologize. I updated those now. Thank you for the catch.

**Review by buckd — APPROVED** [review]

**Pull request merged**